File: code/lt0973.cpp
#include <vector>
#include <utility>
#include <queue>

using namespace std;

class Solution {
public:
    
    struct cmp
    {
        bool operator()(const pair<int, double>& p1, const pair<int, int>& p2)
        {
            return p1.second < p2.second;
        }
    };
    
    vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
        priority_queue<pair<int, double>, vector<pair<int, double>>, cmp> q;
        int size = points.size();
        
        for(int i = 0; i < size; i++)
        {
            pair<int, double> temp = {i, dis(points[i])};
            q.push(temp);
            if(i >= K)
                q.pop();
        }
        
        vector<vector<int>> res = {};
        
        while(!q.empty())
        {
            res.push_back(points[q.top().first]);
            q.pop();
        }
        
        return res;
    }
    
    inline double dis(vector<int>& point)
    {
        return point[0] * point[0] + point[1] * point[1];
    }
};

File: code/lt0973.java
// https://leetcode-cn.com/problems/k-closest-points-to-origin/
// k closest points to origin
class Solution {
    public int[][] kClosest(int[][] points, int K) {
        if(K == 0 || points.length == 0)
            return null;
        
        
        Arrays.sort(points, new Comparator<int[]>(){
           public int compare(int[] p1, int[] p2){
                int d1 = p1[0] * p1[0] + p1[1] * p1[1];
                int d2 = p2[0] * p2[0] + p2[1] * p2[1];
                return d1 - d2;
            } 
        });
        
        int[][] res = new int[K][2];
        
        for(int i = 0; i < K; i++)
            res[i] = points[i];
        
        return res;
    }
}